docs/coding-guidelines/source-generator-guidelines.md

```diff
@@ -0,0 +1,31 @@
+Recommended reading to better understand source generators,
+[Roslyn Source Generators Cookbook](https://github.com/dotnet/roslyn/blob/main/docs/features/source-generators.cookbook.md).
+
+[Project guidance](./project-guidelines.md#directory-layout)
+
+[Packaging guidance](./libraries-packaging.md#analyzers--source-generators)
+
+## Source Generator Best Practices
+
+### DOs
+
+- **DO** generate code that looks as if a developer would write it manually.
+- **DO** emit strings rather than using the Roslyn Syntax API for better performance.
+- **DO** use consistent indentation and formatting in generated code.
+- **DO** generators should use the [`IIncrementalGenerator`](https://learn.microsoft.com/dotnet/api/microsoft.codeanalysis.iincrementalgenerator) interface.
+- **DO** set `<EnforceExtendedAnalyzerRules>true</EnforceExtendedAnalyzerRules>` in the source generator project and then don't use any of the banned APIs it lists.
+- **DO** set `<IsRoslynComponent>true</IsRoslynComponent>` in the source generator project to enable debugging support in Visual Studio.
+- **DO** disable the following Roslyn warning, `RS2008`, in the source generator project (that is, `<NoWarn>$(NoWarn);RS2008</NoWarn>`). The reported issue is handled differently in the runtime repo.
+- **DO** emit diagnostics from a separate [analyzer](https://learn.microsoft.com/visualstudio/extensibility/getting-started-with-roslyn-analyzers). The analyzer and source generator can be in the same assembly.
+- **DO** cache intermediate results to avoid redundant computation.
+- **DO** consider the impact on build time and optimize accordingly.
+- **DO** have separate projects for testing the generator and testing the code generated by the generator.
+- **DO** use the Roslyn Testing SDK to test the generator (and any corresponding analyzers).
+
+### DON'Ts
+
+- **DON'T** use the Roslyn Syntax API for emitting source code.
+- **DON'T** perform expensive operations during the generation process unless absolutely necessary.
+- **DON'T** emit code that introduces runtime dependencies not explicitly referenced by the project.
+- **DON'T** emit code that would trigger compiler warnings in normal usage scenarios.
+- **DON'T** emit diagnostics from the generator itself, emit them from an [analyzer](https://learn.microsoft.com/visualstudio/extensibility/getting-started-with-roslyn-analyzers).
```

Comment:
Add a "don't root symbols and any compilation object in incremental steps"? It's a common mistake.

Reply:
I'm not sure saying "don't root objects" is very helpful. No one wants to do that nor likely does it on purpose, it is a consequence of the programming model that makes it so easy. Perhaps a clearer more actionable statement is "don't store Roslyn instances in long lived fields or collections", is that what you're saying?

Reply:
No (though that's also very good advice), what I was referring to is not using those types in incremental steps, like, having some `IncrementalValuesProvider<ISymbol>` or equivalent. Same with `Compilation`, `SemanticModel`, etc.

All of those will root compilations and also break incrementality in the process, as your entire pipeline will be invalidated up to that point for every single run anyway.

IIRC @Youssef1313 was working on an analyzer for this at some point, but I'm not sure that's available yet 🤔

Reply:
That is a description for how to use Roslyn source generators in general. It is something that should be in the referenced Roslyn Source Generators Cookbook, which should be the standard for how to use the APIs correctly. This is about using them in the runtime and using them correctly is assumed.

Reply:
None of those are "incorrect" uses. It is recommended to use incremental, but still possible not to. Expensive work doesn't break any invariants, it just makes for a bad experience. Formating with strings isn't required, but using the syntax is completely okay.

Using the API incorrectly means the benefit to incrementality is broken and wrong. These are not the same thing.